src/OpticalFlowEstimator.cpp

```diff
@@ -586,15 +617,388 @@ bool OpticalFlowEstimator::findAverageVector(
+    mean_and_var(no_outlier_deltas_y, filtered_u_y, filtered_var_y);
+
+    double filtered_covariance = 0.0;
+    covariance(no_outlier_deltas_x,
+               no_outlier_deltas_y,
+               filtered_u_x,
+               filtered_u_y,
+               filtered_covariance);
+
+    Eigen::Matrix2d filtered_covariance_matrix = Eigen::Matrix2d::Zero();
+    filtered_covariance_matrix(0, 0) = filtered_var_x;
+    filtered_covariance_matrix(0, 1) = filtered_covariance;
+    filtered_covariance_matrix(1, 0) = filtered_covariance;
+    filtered_covariance_matrix(1, 1) = filtered_var_y;
+
+    Eigen::SelfAdjointEigenSolver<Eigen::Matrix2d> filtered_covariance_eigen(filtered_covariance_matrix);
+
+    bool filtered_variance_accepted =
+        (filtered_covariance_eigen.eigenvalues()[0] <= flow_estimator_settings_.max_filtered_variance)
+        & (filtered_covariance_eigen.eigenvalues()[1] <= flow_estimator_settings_.max_filtered_variance);
+
+    if(!filtered_variance_accepted) {
+        ROS_WARN_STREAM("iarc7_vision: Flow filtered variance out of bounds. Max: "
+                        << flow_estimator_settings_.max_filtered_variance
+                        << " Actual: "
+                        << filtered_covariance_eigen.eigenvalues()[0]
+                        << ", "
+                        << filtered_covariance_eigen.eigenvalues()[1]);
+    }
+
+    iarc7_msgs::FlowQuality flow_quality_msg;
+    flow_quality_msg.header.stamp = time;
+
+    flow_quality_msg.num_starting_vectors = dx.size()
+                                            + rejection_region_dx.size()
+                                            + roomba_region_dx.size();
+    flow_quality_msg.num_rejection_region_vectors = rejection_region_dx.size();
+    flow_quality_msg.num_roomba_region_vectors = roomba_region_dx.size();
+    flow_quality_msg.num_exceeded_element_var_vectors
+        = dx.size() - no_outlier_deltas_x.size();
+    flow_quality_msg.num_accepted_vectors = no_outlier_deltas_x.size();
+
+    flow_quality_msg.sample_std_dev_eigen_values.x
+        = std::sqrt(sample_covariance_eigen.eigenvalues()[0]);
+    flow_quality_msg.sample_std_dev_eigen_values.y
+        = std::sqrt(sample_covariance_eigen.eigenvalues()[1]);
+    flow_quality_msg.sample_average.x = sample_u_x;
+    flow_quality_msg.sample_average.y = sample_u_y;
+
+    flow_quality_msg.filtered_std_dev_eigen_values.x
+        = std::sqrt(filtered_covariance_eigen.eigenvalues()[0]);
+    flow_quality_msg.filtered_std_dev_eigen_values.y
+        = std::sqrt(filtered_covariance_eigen.eigenvalues()[1]);
+    flow_quality_msg.filtered_average.x = filtered_u_x;
+    flow_quality_msg.filtered_average.y = filtered_u_y;
+
+    flow_quality_msg.diff_filtered_and_sample_avg.x
+        = filtered_u_x - sample_u_x;
+    flow_quality_msg.diff_filtered_and_sample_avg.y
+        = filtered_u_y - sample_u_y;
+
+    debug_flow_quality_pub_.publish(flow_quality_msg);
+
+    if (debug && debug_settings_.debug_hist) {
```

Comment:
lol

Reply:
I'd rather not, it requires a ton of parameters that are local to this function. Given that there's a good chance this will require modification in the future during a testing session that may include the addition of parameters I don't think moving the histogram function to its own function is a good idea.

Reply:
It's 200 lines in the middle of the logical flow of this function, it needs to move